types/mithril/index.d.ts

```diff
@@ -204,7 +201,7 @@ declare namespace Mithril {
     type ChildArrayOrPrimitive = ChildArray | string | number | boolean;
 
     /** Virtual DOM nodes, or vnodes, are Javascript objects that represent an element (or parts of the DOM). */
-    interface Vnode<Attrs = {}, State extends Lifecycle<Attrs, State> = {}> {
```

Comment:
This is actually a necessary invariant and is technically insufficient as well (as in we need to revise the typings a little) - `vnode.state` is directly used internally for lifecycle hooks at the component-level. ([Example](https://github.com/MithrilJS/mithril.js/blob/773e570b86618be5f9de5b254c8c42e5beabf9d5/render/render.js#L139-L158))

Have you tested for code like this?

```ts
// @ts-expect-error The real argument is a vnode, not a literal DOM element.
const Foo: m.Component<{}, {oncreate(dom: HTMLElement): void}> = {
  oninit: v => {
    // @ts-expect-error The real argument is a vnode, not a literal DOM element.
    v.state.oncreate = (dom: HTMLElement) => { dom.classList.add("ready") }
  },

  view: () => "children",
}
```

Reply:
I'd be open to looking into alternatives, but we do need a self-referential constraint. It *may* be doable using `this` types, but we'd have to half rewrite the typings, which would be a relatively breaking change.

Reply:
I don't know mithril, so I can't see from the types alone how a concrete state interface like `interface State { count: number }` is an instance of `LifeCycle`, so I will just silence the errors in a different PR.